scheduler/test/cook/test/kubernetes/controller.clj

```diff
@@ -38,39 +38,62 @@
       (is (= :cook-expected-state/killed (do-process :cook-expected-state/killed :pod/waiting)))
 
       (is (nil? (do-process :cook-expected-state/running :missing)))
-      (is (nil? (do-process :cook-expected-state/running :pod/succeeded)))
-      (is (nil? (do-process :cook-expected-state/running :pod/failed)))
+      (is (= :cook-expected-state/completed (do-process :cook-expected-state/running :pod/succeeded)))
+      (is (= :cook-expected-state/completed (do-process :cook-expected-state/running :pod/failed)))
       (is (= :cook-expected-state/running (do-process :cook-expected-state/running :pod/running)))
       (is (= :cook-expected-state/completed (do-process :cook-expected-state/running :pod/unknown)))
       (is (= :cook-expected-state/completed (do-process :cook-expected-state/running :pod/waiting)))
 
       (is (= :cook-expected-state/starting (do-process :cook-expected-state/starting :missing)))
-      (is (nil? (do-process :cook-expected-state/starting :pod/succeeded)))
-      (is (nil? (do-process :cook-expected-state/starting :pod/failed)))
+      (is (= :cook-expected-state/completed (do-process :cook-expected-state/starting :pod/succeeded)))
+      (is (= :cook-expected-state/completed (do-process :cook-expected-state/starting :pod/failed)))
       (is (= :cook-expected-state/running (do-process :cook-expected-state/starting :pod/running)))
       (is (= :cook-expected-state/completed (do-process :cook-expected-state/starting :pod/unknown)))
       (is (= :cook-expected-state/starting (do-process :cook-expected-state/starting :pod/waiting)))
 
       (is (nil? (do-process :missing :missing)))
       (is (nil? (do-process :missing :pod/succeeded)))
       (is (nil? (do-process :missing :pod/failed)))
-      (is (nil? (do-process :missing :pod/running)))
+      (is (= :missing (do-process :missing :pod/running)))
       (is (nil? (do-process :missing :pod/unknown)))
-      (is (nil? (do-process :missing :pod/waiting))))))
+      (is (= :missing (do-process :missing :pod/waiting))))))
+
+(deftest test-completion-protocol
+  (let [name "TestPodName"
+        do-process (fn [cook-expected-state k8s-actual-state]
+                     (let [cook-expected-state-map
+                           (atom {name {:cook-expected-state cook-expected-state}})]
+                       (controller/process
+                         {:api-client nil
+                          :cook-expected-state-map cook-expected-state-map
+                          :k8s-actual-state-map (atom {name {:synthesized-state {:state k8s-actual-state} :pod nil}})}
+                         name)
+                       (:cook-expected-state (get @cook-expected-state-map name {}))))
+        count-delete-pod (atom 0)]
+    (with-redefs [controller/delete-pod  (fn [_ cook-expected-state-dict _] (swap! count-delete-pod inc) cook-expected-state-dict)
+                  controller/log-weird-state (fn [_ _ _ _] :illegal_return_value_should_be_unused)
```

Comment:
Why do we need to `with-redefs` the `log-weird-state` call?

Reply:
Here it doesn't make any sense, it just avoids spamming the log.